### PR TITLE
Standardize creation of FigureManager from a given FigureCanvas class.

### PR DESCRIPTION
**Pull request by anntzer**
The `new_manager` classmethod may appear to belong more naturally to the
FigureManager class rather than the FigureCanvas class, but putting it
on FigureCanvas has multiple advantages:

- One may want to create managers at times where all one has is a
  FigureCanvas instance, which may not even have a corresponding manager
  (e.g. `subplot_tool`).
- A given FigureManager class can be associated with many different
  FigureCanvas classes (indeed, FigureManagerQT can manage both
  FigureCanvasQTAgg and FigureCanvasQTCairo and also the mplcairo Qt
  canvas classes), whereas we don't have multiple FigureManager classes
  for a given FigureCanvas class.

Still, put the *actual* logic in FigureManager.create_with_canvas and
access it via some indirection, as suggested by timhoffm (#22895).

See discussion at #18854 and #22895.

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Review by tacaswell — APPROVED** [review]

File: lib/matplotlib/backend_bases.py [hunk 1]
```diff
@@ -1662,6 +1669,13 @@ def _fix_ipython_backend2gui(cls):
             if _is_non_interactive_terminal_ipython(ip):
                 ip.enable_gui(backend2gui_rif)
 
+    @classmethod
+    def new_manager(cls, figure, num):
+        """
+        Create a new figure manager for *figure*, using this canvas class.
+        """
```
Comment (timhoffm) [inline lib/matplotlib/backend_bases.py:1676]:
```suggestion

        Notes:
            This should usually not be reimplemented in subclasses. If custom
            manager creation logic is needed, please reimplement
            ``FigureManager.create_with_canvas``.
        """
```

File: lib/matplotlib/backend_bases.py [hunk 2]
```diff
@@ -1580,6 +1580,13 @@ class FigureCanvasBase:
     # interactive framework is required, or None otherwise.
     required_interactive_framework = None
 
+    # The manager class instantiated by new_manager.
+    # (This is defined as a classproperty because the manager class is
+    # currently defined *after* the canvas class, but one could also assign
+    # ``FigureCanvasBase.manager_class = FigureManagerBase``
+    # after defining both classes.)
+    manager_class = _api.classproperty(lambda cls: FigureManagerBase)
```
Comment (timhoffm) [inline lib/matplotlib/backend_bases.py:1588]: Can the manager class be private?

Comment (anntzer) [reply lib/matplotlib/backend_bases.py:1588]: This is intended to be a public customization point for third-party backends, and should therefore be public, IMO.

File: lib/matplotlib/backend_bases.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/backend_bases.py:1682]: done.

File: lib/matplotlib/backend_bases.py (hunk 2 above)
Comment (timhoffm) [reply lib/matplotlib/backend_bases.py:1588]: Technically, we'd want what other languages descripbe as *protected*: Accessible for subclasses, but not for users. I think this is still best translated to python by adding a leading underscore.

Comment (anntzer) [reply lib/matplotlib/backend_bases.py:1588]: Unfortunately, all we have is whether to put an underscore or not, and I would rather have that mean "do we allow ourselves to change this API without deprecation or not"; in this specific case the answer is clearly "not" as I think this party backends should indeed use it. (It's also not the end of the world if end users access this attribute, I think?)

Comment (timhoffm) [reply lib/matplotlib/backend_bases.py:1588]: Fair.

**Review by timhoffm — APPROVED** [review]

**Pull request merged**